File: backend/app/schemas/agent_run.py
from __future__ import annotations

from datetime import datetime
import uuid

from pydantic import BaseModel, Field


class AgentRunCreate(BaseModel):
    task_type: str = Field(default="candidate_next_step", max_length=80)
    target_type: str = Field(default="analysis", max_length=80)
    target_id: uuid.UUID
    task_prompt: str | None = Field(default=None, max_length=2000)
    context: dict | None = None
    run_async: bool = True


class AgentRunStepItem(BaseModel):
    id: uuid.UUID
    step_index: int
    agent_name: str
    status: str
    input_packet: dict | None = None
    output: dict | None = None
    summary: str | None = None
    error: str | None = None
    started_at: datetime
    completed_at: datetime | None = None

    model_config = {"from_attributes": True}


class AgentRunItem(BaseModel):
    id: uuid.UUID
    task_type: str
    target_type: str
    target_id: uuid.UUID
    task_prompt: str | None = None
    status: str
    current_agent: str | None = None
    result: dict | None = None
    error: str | None = None
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None = None
    steps: list[AgentRunStepItem] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class AgentRunListResponse(BaseModel):
    items: list[AgentRunItem]

File: frontend/src/types/agentRun.ts
export type AgentRunStatus = 'pending' | 'running' | 'completed' | 'failed';
export type AgentRunStepStatus = 'running' | 'completed' | 'failed';

export interface AgentRunStepItem {
  id: string;
  step_index: number;
  agent_name: string;
  status: AgentRunStepStatus;
  input_packet?: Record<string, unknown> | null;
  output?: Record<string, unknown> | null;
  summary?: string | null;
  error?: string | null;
  started_at: string;
  completed_at?: string | null;
}

export interface AgentRunResult {
  summary?: string | null;
  recommended_next_step?: string | null;
  safe_to_auto_apply?: boolean;
  human_review_points?: string[];
  agent_outputs?: Record<string, Record<string, unknown>>;
  [key: string]: unknown;
}

export interface AgentRunItem {
  id: string;
  task_type: string;
  target_type: string;
  target_id: string;
  task_prompt?: string | null;
  status: AgentRunStatus;
  current_agent?: string | null;
  result?: AgentRunResult | null;
  error?: string | null;
  created_at: string;
  updated_at: string;
  completed_at?: string | null;
  steps: AgentRunStepItem[];
}

export interface AgentRunListResponse {
  items: AgentRunItem[];
}

export interface AgentRunCreatePayload {
  task_type?: string;
  target_type: 'analysis';
  target_id: string;
  task_prompt?: string;
  context?: Record<string, unknown>;
  run_async?: boolean;
}
